FIAT/brezzi_douglas_marini_cube.py

```diff
@@ -0,0 +1,229 @@
+        self._degree = degree
+        self.flat_el = flat_el
+
+    def degree(self):
+        return self._degree
+
+    def get_nodal_basis(self):
+        raise NotImplementedError("get_nodal_basis not implemented for bdmce")
+
+    def get_dual_set(self):
+        raise NotImplementedError("get_dual_set is not implemented for bdmce")
+
+    def get_coeffs(self):
+        raise NotImplementedError("get_coeffs not implemented for bdmce")
+
+    def tabulate(self, order, points, entity=None):
+
+        if entity is None:
+            entity = (self.ref_el.get_dimension(), 0)
+
+        entity_dim, entity_id = entity
+        transform = self.ref_el.get_entity_transform(entity_dim, entity_id)
+        points = list(map(transform, points))
+
+        phivals = {}
+
+        for o in range(order+1):
+            alphas = mis(2, o)
+            for alpha in alphas:
+                try:
+                    polynomials = self.basis[alpha]
+                except KeyError:
+                    polynomials = diff(self.basis[(0, 0)], *zip(variables, alpha))
+                    self.basis[alpha] = polynomials
+                T = np.zeros((len(polynomials[:, 0]), 2, len(points)))
+                for i in range(len(points)):
+                    subs = {v: points[i][k] for k, v in enumerate(variables[:2])}
+                    for j, f in enumerate(polynomials[:, 0]):
+                        T[j, 0, i] = f.evalf(subs=subs)
+                    for j, f in enumerate(polynomials[:, 1]):
+                        T[j, 1, i] = f.evalf(subs=subs)
+                phivals[alpha] = T
+
+        return phivals
+
+    def entity_dofs(self):
+        """Return the map of topological entities to degrees of
+        freedom for the finite element."""
+        return self.entity_ids
+
+    def entity_closure_dofs(self):
+        """Return the map of topological entities to degrees of
+        freedom on the closure of those entities for the finite element."""
+        return self.entity_closure_ids
+
+    def value_shape(self):
+        return (2,)
+
+    def dmats(self):
+        raise NotImplementedError
+
+    def get_num_members(self, arg):
+        raise NotImplementedError
+
+    def space_dimension(self):
+        return int(len(self.basis[(0, 0)])/2)
+
+
+def e_lambda_1_2d(deg, dx, dy, x_mid, y_mid):
+    EL = tuple([(0, -leg(j, y_mid)*dx[0]) for j in range(deg)] +
+               [(-leg(deg-1, y_mid)*dy[0]*dy[1]/(deg+1), -leg(deg, y_mid)*dx[0])] +
+               [(0, -leg(j, y_mid)*dx[1]) for j in range(deg)] +
+               [(leg(deg-1, y_mid)*dy[0]*dy[1]/(deg+1), -leg(deg, y_mid)*dx[1])] +
+               [(-leg(j, x_mid)*dy[0], 0) for j in range(deg)] +
+               [(-leg(deg, x_mid)*dy[0], -leg(deg-1, x_mid)*dx[0]*dx[1]/(deg+1))] +
+               [(-leg(j, x_mid)*dy[1], 0) for j in range(deg)] +
+               [(-leg(deg, x_mid)*dy[1], leg(deg-1, x_mid)*dx[0]*dx[1]/(deg+1))])
+
```

Comment:
My understanding of the issues @pefarrell was reporting with the lowest-order spaces is that these basis functions aren't quite right.

When taking the 2D curl of each basis function (or the divergence for BDMCF) it should return a polynomial of `deg-1`. 

Inspection of these basis functions shows that this is correct for the functions such as
```[(0, -leg(j, y_mid)*dx[0]) for j in range(deg)]```
but not quite for those like
```[(-leg(deg-1, y_mid)*dy[0]*dy[1]/(deg+1), -leg(deg, y_mid)*dx[0])]```.

I did a bit of maths and I think that one component of these needs modifying by a coefficient. Running the `bdmc_riesz` test in the `bdmc` branch of Firedrake, I got a convergence rate of `deg+1` when using the following functions (which ensure that the curl/div of the basis functions is of the correct polynomial order):
```
from sympy import binomial
coeff = binomial(2*deg,deg) / ((deg+1)*binomial(2*(deg-1), deg-1))
EL = tuple([(0, -leg(j, y_mid)*dx[0]) for j in range(deg)] +
           [(coeff*-leg(deg-1, y_mid)*dy[0]*dy[1]/(deg+1), -leg(deg, y_mid)*dx[0])] +
           [(0, -leg(j, y_mid)*dx[1]) for j in range(deg)] +
           [(coeff*leg(deg-1, y_mid)*dy[0]*dy[1]/(deg+1), -leg(deg, y_mid)*dx[1])] +
           [(-leg(j, x_mid)*dy[0], 0) for j in range(deg)] +
           [(-leg(deg, x_mid)*dy[0], coeff*-leg(deg-1, x_mid)*dx[0]*dx[1]/(deg+1))] +
           [(-leg(j, x_mid)*dy[1], 0) for j in range(deg)] +
           [(-leg(deg, x_mid)*dy[1], coeff*leg(deg-1, x_mid)*dx[0]*dx[1]/(deg+1))])
```